source/val/validate_memory.cpp

```diff
@@ -323,6 +324,34 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
              << "Classes: Workgroup, CrossWorkgroup, Private, and Function";
     }
   }
+
+  // Vulkan spec 14.5.2
+  if (spvIsVulkanEnv(_.context()->target_env) &&
+      storage_class == SpvStorageClassUniformConstant) {
+    auto variable_type = _.FindDef(result_type->GetOperandAs<uint32_t>(2));
+    auto variable_type_opcode = variable_type->opcode();
+
+    // If the variable is actually an array extract the element type.
+    if (variable_type_opcode == SpvOpTypeArray ||
+        variable_type_opcode == SpvOpTypeRuntimeArray) {
+      variable_type = _.FindDef(variable_type->GetOperandAs<uint32_t>(1));
+      variable_type_opcode = variable_type->opcode();
+    }
+
+    switch (variable_type_opcode) {
+      case SpvOpTypeImage:
+      case SpvOpTypeSampler:
+      case SpvOpTypeSampledImage:
+        break;
+      default:
+        return _.diag(SPV_ERROR_INVALID_ID, inst)
+               << "Variables identified with the UniformConstant storage class "
```

Comment:
I generally like to mention this is a Vulkan specific check so the user knows where to look.

Reply:
Updated the comment at the beginning of this block to be more explicit.

Reply:
I meant this error message. The user should have to read this code to know its a Vulkan-specific check.

Reply:
my bad, I have added a reference to the spec in the error message